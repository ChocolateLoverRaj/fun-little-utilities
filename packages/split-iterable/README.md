# Split Iterable

Split iterable into multiple chunks

## Usage

### `splitIterable`

```javascript
const { splitIterable } = require('split-iterable')

expect(Array.from(
  splitIterable([0, 'x', 1, 2, 'x', 3, 4, 5])
)).toEqual([
  [0],
  [1, 2],
  [3, 4, 5]
])
```

### `splitAsyncIterable`

```javascript
const { splitAsyncIterable } = require('split-iterable')

const iterable = {
  async * [Symbol.asyncIterator] () {
    yield * [0, 'x', 1, 2, 'x', 3, 4, 5]
  }
}

async function toArray (iterable) {
  let array = []

  for await (const item of iterable) {
    array.push(item)
  }

  return array
}

expect(await toArray(
  splitAsyncIterable(iterable)
)).toEqual([
  [0],
  [1, 2],
  [3, 4, 5]
])
```

## License

[MIT](https://git.io/fxKXN) © [Hoàng Văn Khải](https://github.com/KSXGitHub)
