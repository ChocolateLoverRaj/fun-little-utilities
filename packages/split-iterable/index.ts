interface SepFunc<Item> {
  (item: Item): boolean
}

export function* splitIterable<Item>(iterable: Iterable<Item>, sep: SepFunc<Item>) {
  let store = Array<Item>()

  for (const item of iterable) {
    if (sep(item)) {
      yield store
      store = []
    } else {
      store.push(item)
    }
  }

  yield store
}

type MaybeAsyncIterable<Item> = Iterable<Item> | AsyncIterable<Item>

interface AsyncSepFunc<Item> {
  (item: Item): boolean | Promise<boolean>
}

export async function* splitAsyncIterable<Item>(
  iterable: MaybeAsyncIterable<Item>,
  sep: AsyncSepFunc<Item>,
) {
  let store = Array<Item>()

  for await (const item of iterable) {
    if (await sep(item)) {
      yield store
      store = []
    } else {
      store.push(item)
    }
  }

  yield store
}
