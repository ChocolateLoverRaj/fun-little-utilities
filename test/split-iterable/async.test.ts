import { splitAsyncIterable } from 'split-iterable'

import {
  SPLITTER,
  sep,
  asyncSep,
  args2arr,
  getArrayFromAsyncIterable,
  asyncIterate,
  assertResemble,
  assertNoSplitter
} from './utils'

describe('with sync iterator and sync separator', () => {
  const get =
    <X> (iter: Iterable<X>, sep: (x: X) => boolean) =>
      getArrayFromAsyncIterable(splitAsyncIterable(iter, sep))

  it('with empty array', async () => {
    expect(await get([], sep)).toEqual([[]])
  })

  describe('with iterable that lacks splitter', () => {
    it('makes exactly one chunk', async () => {
      expect(await get('abcdef', sep)).toHaveProperty('length', 1)
    })

    it('resembles original iterable', async () => {
      assertResemble('abcdef', await get('abcdef', sep))
    })

    it('does not contain splitter elements', async () => {
      assertNoSplitter(await get('abcdef', sep))
    })

    it('matches snapshot', async () => {
      expect(await get('abcdef', sep)).toMatchSnapshot()
    })
  })

  describe('with iterable that has splitters', () => {
    const iterable = args2arr(0, 1, 2, SPLITTER, 3, 4, SPLITTER, 5)

    it('makes multiple chunks', async () => {
      expect(await get(iterable, sep)).toHaveProperty('length', 3)
    })

    it('resembles original iterable', async () => {
      assertResemble(iterable, await get(iterable, sep) as any)
    })

    it('does not contain splitter elements', async () => {
      assertNoSplitter(await get(iterable, sep))
    })

    it('matches snapshot', async () => {
      expect(await get(iterable, sep)).toMatchSnapshot()
    })
  })

  describe('with iterable that has trailing splitters', () => {
    const iterable = args2arr(SPLITTER, 0, 1, 2, SPLITTER, 3, 4, SPLITTER, 5, SPLITTER)

    it('makes multiple chunks', async () => {
      expect(await get(iterable, sep)).toHaveProperty('length', 3 + 2)
    })

    it('first chunk is empty', async () => {
      expect((await get(iterable, sep))[0]).toEqual([])
    })

    it('last chunk is empty', async () => {
      expect((await get(iterable, sep)).slice(-1)[0]).toEqual([])
    })

    it('resembles original iterable', async () => {
      assertResemble(iterable, await get(iterable, sep) as any)
    })

    it('does not contain splitter elements', async () => {
      assertNoSplitter(await get(iterable, sep))
    })

    it('matches snapshot', async () => {
      expect(await get(iterable, sep)).toMatchSnapshot()
    })
  })
})

describe('with async iterator and sync separator', () => {
  const get =
    <X> (iter: Iterable<X>, sep: (x: X) => boolean) =>
      getArrayFromAsyncIterable(splitAsyncIterable(asyncIterate(iter), sep))

  it('with empty array', async () => {
    expect(await get([], sep)).toEqual([[]])
  })

  describe('with iterable that lacks splitter', () => {
    it('makes exactly one chunk', async () => {
      expect(await get('abcdef', sep)).toHaveProperty('length', 1)
    })

    it('resembles original iterable', async () => {
      assertResemble('abcdef', await get('abcdef', sep))
    })

    it('does not contain splitter elements', async () => {
      assertNoSplitter(await get('abcdef', sep))
    })

    it('matches snapshot', async () => {
      expect(await get('abcdef', sep)).toMatchSnapshot()
    })
  })

  describe('with iterable that has splitters', () => {
    const iterable = args2arr(0, 1, 2, SPLITTER, 3, 4, SPLITTER, 5)

    it('makes multiple chunks', async () => {
      expect(await get(iterable, sep)).toHaveProperty('length', 3)
    })

    it('resembles original iterable', async () => {
      assertResemble(iterable, await get(iterable, sep) as any)
    })

    it('does not contain splitter elements', async () => {
      assertNoSplitter(await get(iterable, sep))
    })

    it('matches snapshot', async () => {
      expect(await get(iterable, sep)).toMatchSnapshot()
    })
  })

  describe('with iterable that has trailing splitters', () => {
    const iterable = args2arr(SPLITTER, 0, 1, 2, SPLITTER, 3, 4, SPLITTER, 5, SPLITTER)

    it('makes multiple chunks', async () => {
      expect(await get(iterable, sep)).toHaveProperty('length', 3 + 2)
    })

    it('first chunk is empty', async () => {
      expect((await get(iterable, sep))[0]).toEqual([])
    })

    it('last chunk is empty', async () => {
      expect((await get(iterable, sep)).slice(-1)[0]).toEqual([])
    })

    it('resembles original iterable', async () => {
      assertResemble(iterable, await get(iterable, sep) as any)
    })

    it('does not contain splitter elements', async () => {
      assertNoSplitter(await get(iterable, sep))
    })

    it('matches snapshot', async () => {
      expect(await get(iterable, sep)).toMatchSnapshot()
    })
  })
})

describe('with sync iterator and async separator', () => {
  const get =
    <X> (iter: Iterable<X>, asyncSep: (x: X) => Promise<boolean>) =>
      getArrayFromAsyncIterable(splitAsyncIterable(iter, asyncSep))

  it('with empty array', async () => {
    expect(await get([], asyncSep)).toEqual([[]])
  })

  describe('with iterable that lacks splitter', () => {
    it('makes exactly one chunk', async () => {
      expect(await get('abcdef', asyncSep)).toHaveProperty('length', 1)
    })

    it('resembles original iterable', async () => {
      assertResemble('abcdef', await get('abcdef', asyncSep))
    })

    it('does not contain splitter elements', async () => {
      assertNoSplitter(await get('abcdef', asyncSep))
    })

    it('matches snapshot', async () => {
      expect(await get('abcdef', asyncSep)).toMatchSnapshot()
    })
  })

  describe('with iterable that has splitters', () => {
    const iterable = args2arr(0, 1, 2, SPLITTER, 3, 4, SPLITTER, 5)

    it('makes multiple chunks', async () => {
      expect(await get(iterable, asyncSep)).toHaveProperty('length', 3)
    })

    it('resembles original iterable', async () => {
      assertResemble(iterable, await get(iterable, asyncSep) as any)
    })

    it('does not contain splitter elements', async () => {
      assertNoSplitter(await get(iterable, asyncSep))
    })

    it('matches snapshot', async () => {
      expect(await get(iterable, asyncSep)).toMatchSnapshot()
    })
  })

  describe('with iterable that has trailing splitters', () => {
    const iterable = args2arr(SPLITTER, 0, 1, 2, SPLITTER, 3, 4, SPLITTER, 5, SPLITTER)

    it('makes multiple chunks', async () => {
      expect(await get(iterable, asyncSep)).toHaveProperty('length', 3 + 2)
    })

    it('first chunk is empty', async () => {
      expect((await get(iterable, asyncSep))[0]).toEqual([])
    })

    it('last chunk is empty', async () => {
      expect((await get(iterable, asyncSep)).slice(-1)[0]).toEqual([])
    })

    it('resembles original iterable', async () => {
      assertResemble(iterable, await get(iterable, asyncSep) as any)
    })

    it('does not contain splitter elements', async () => {
      assertNoSplitter(await get(iterable, asyncSep))
    })

    it('matches snapshot', async () => {
      expect(await get(iterable, asyncSep)).toMatchSnapshot()
    })
  })
})

describe('with async iterator and async separator', () => {
  const get =
    <X> (iter: Iterable<X>, asyncSep: (x: X) => Promise<boolean>) =>
      getArrayFromAsyncIterable(splitAsyncIterable(asyncIterate(iter), asyncSep))

  it('with empty array', async () => {
    expect(await get([], asyncSep)).toEqual([[]])
  })

  describe('with iterable that lacks splitter', () => {
    it('makes exactly one chunk', async () => {
      expect(await get('abcdef', asyncSep)).toHaveProperty('length', 1)
    })

    it('resembles original iterable', async () => {
      assertResemble('abcdef', await get('abcdef', asyncSep))
    })

    it('does not contain splitter elements', async () => {
      assertNoSplitter(await get('abcdef', asyncSep))
    })

    it('matches snapshot', async () => {
      expect(await get('abcdef', asyncSep)).toMatchSnapshot()
    })
  })

  describe('with iterable that has splitters', () => {
    const iterable = args2arr(0, 1, 2, SPLITTER, 3, 4, SPLITTER, 5)

    it('makes multiple chunks', async () => {
      expect(await get(iterable, asyncSep)).toHaveProperty('length', 3)
    })

    it('resembles original iterable', async () => {
      assertResemble(iterable, await get(iterable, asyncSep) as any)
    })

    it('does not contain splitter elements', async () => {
      assertNoSplitter(await get(iterable, asyncSep))
    })

    it('matches snapshot', async () => {
      expect(await get(iterable, asyncSep)).toMatchSnapshot()
    })
  })

  describe('with iterable that has trailing splitters', () => {
    const iterable = args2arr(SPLITTER, 0, 1, 2, SPLITTER, 3, 4, SPLITTER, 5, SPLITTER)

    it('makes multiple chunks', async () => {
      expect(await get(iterable, asyncSep)).toHaveProperty('length', 3 + 2)
    })

    it('first chunk is empty', async () => {
      expect((await get(iterable, asyncSep))[0]).toEqual([])
    })

    it('last chunk is empty', async () => {
      expect((await get(iterable, asyncSep)).slice(-1)[0]).toEqual([])
    })

    it('resembles original iterable', async () => {
      assertResemble(iterable, await get(iterable, asyncSep) as any)
    })

    it('does not contain splitter elements', async () => {
      assertNoSplitter(await get(iterable, asyncSep))
    })

    it('matches snapshot', async () => {
      expect(await get(iterable, asyncSep)).toMatchSnapshot()
    })
  })
})
