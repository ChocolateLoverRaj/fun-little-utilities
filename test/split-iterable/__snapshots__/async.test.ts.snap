// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with async iterator and async separator with iterable that has splitters matches snapshot 1`] = `
Array [
  Array [
    0,
    1,
    2,
  ],
  Array [
    3,
    4,
  ],
  Array [
    5,
  ],
]
`;

exports[`with async iterator and async separator with iterable that has trailing splitters matches snapshot 1`] = `
Array [
  Array [],
  Array [
    0,
    1,
    2,
  ],
  Array [
    3,
    4,
  ],
  Array [
    5,
  ],
  Array [],
]
`;

exports[`with async iterator and async separator with iterable that lacks splitter matches snapshot 1`] = `
Array [
  Array [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
  ],
]
`;

exports[`with async iterator and sync separator with iterable that has splitters matches snapshot 1`] = `
Array [
  Array [
    0,
    1,
    2,
  ],
  Array [
    3,
    4,
  ],
  Array [
    5,
  ],
]
`;

exports[`with async iterator and sync separator with iterable that has trailing splitters matches snapshot 1`] = `
Array [
  Array [],
  Array [
    0,
    1,
    2,
  ],
  Array [
    3,
    4,
  ],
  Array [
    5,
  ],
  Array [],
]
`;

exports[`with async iterator and sync separator with iterable that lacks splitter matches snapshot 1`] = `
Array [
  Array [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
  ],
]
`;

exports[`with sync iterator and async separator with iterable that has splitters matches snapshot 1`] = `
Array [
  Array [
    0,
    1,
    2,
  ],
  Array [
    3,
    4,
  ],
  Array [
    5,
  ],
]
`;

exports[`with sync iterator and async separator with iterable that has trailing splitters matches snapshot 1`] = `
Array [
  Array [],
  Array [
    0,
    1,
    2,
  ],
  Array [
    3,
    4,
  ],
  Array [
    5,
  ],
  Array [],
]
`;

exports[`with sync iterator and async separator with iterable that lacks splitter matches snapshot 1`] = `
Array [
  Array [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
  ],
]
`;

exports[`with sync iterator and sync separator with iterable that has splitters matches snapshot 1`] = `
Array [
  Array [
    0,
    1,
    2,
  ],
  Array [
    3,
    4,
  ],
  Array [
    5,
  ],
]
`;

exports[`with sync iterator and sync separator with iterable that has trailing splitters matches snapshot 1`] = `
Array [
  Array [],
  Array [
    0,
    1,
    2,
  ],
  Array [
    3,
    4,
  ],
  Array [
    5,
  ],
  Array [],
]
`;

exports[`with sync iterator and sync separator with iterable that lacks splitter matches snapshot 1`] = `
Array [
  Array [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
  ],
]
`;
